src/Shared/AssemblyNameExtension.cs

```diff
@@ -186,7 +186,23 @@ internal static AssemblyNameExtension GetAssemblyNameEx(string path)
 #if !FEATURE_ASSEMBLYLOADCONTEXT
             try
             {
-                assemblyName = AssemblyName.GetAssemblyName(path);
+                try
+                {
+                    assemblyName = AssemblyName.GetAssemblyName(path);
+                }
+                catch (OutOfMemoryException)
+                {
+                    // Force a full GC and try one more time
+                    // https://github.com/microsoft/msbuild/issues/5150
+                    GC.GetTotalMemory(forceFullCollection: true);
+#if NET35
+                    GC.Collect(GC.MaxGeneration, GCCollectionMode.Forced);
+                    GC.WaitForFullGCComplete();
```

Comment:
`WaitForFullGCComplete` is not doing what it looks like it should be doing. If you want a "wait for all GC related activities are done" call, `WaitForPendingFinalizers` would be a better choice here.

Reply:
No objection to changing this, but keep in mind the behavior of the other section is more important (this section is only for .NET 3.5).

Reply:
➡️ `GC.GetTotalMemory` calls `WaitForPendingFinalizers`+`Collect` in a loop. The code after the call to `GetTotalMemory` is intended to cover the compacting operation.